
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Hierarchical TAD &#8212; TADLib 0.4.0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Hierarchical TAD Identification" href="hitad_api.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">TADLib</a></h1>



<p class="blurb">A Library to Explore Chromatin Interaction Patterns for Topologically Associating Domains</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=XiaoTaoWang&repo=TADLib&type=star&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Install TADLib</a></li>
<li class="toctree-l1"><a class="reference internal" href="calfea.html">Aggregation Preference</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Hierarchical TAD</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#snapshot">Snapshot</a></li>
<li class="toctree-l2"><a class="reference internal" href="#tutorial">Tutorial</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#hierarchical-tad-identification">Hierarchical TAD Identification</a></li>
<li class="toctree-l3"><a class="reference internal" href="#visualization">Visualization</a></li>
<li class="toctree-l3"><a class="reference internal" href="#output-di-track">Output DI track</a></li>
<li class="toctree-l3"><a class="reference internal" href="#domain-alignment">Domain Alignment</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id1">API Documentation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="hitad_api.html">Hierarchical TAD Identification</a></li>
<li class="toctree-l3"><a class="reference internal" href="hitad_api.html#domain-loading-and-aligning">Domain Loading and Aligning</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Release Notes</a></li>
</ul>


<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="hierarchical-tad">
<h1>Hierarchical TAD<a class="headerlink" href="#hierarchical-tad" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>HiTAD is a method to detect hierarchical TADs, including TADs, sub-TADs and smaller domains.
Except local insulations, HiTAD further constrains TADs as the optimal domains to globally
separate intra-chromosomal interactions. Under objective functions derived from chromatin
interactions, HiTAD adopts an iterative optimization procedure to detect hierarchical TADs.
HiTAD performs well in domain sensitivity, replicate reproducibility and inter cell-type
conservation. Application to human and mouse cell types (including both traditional Hi-C
and in situ Hi-C data sets) reveals that there exist common change types for hierarchical
TADs, which are involved in shaping higher-order compartment, replication timing and
transcriptional regulation.</p>
</div>
<div class="section" id="snapshot">
<h2>Snapshot<a class="headerlink" href="#snapshot" title="Permalink to this headline">¶</a></h2>
<p>This tutorial will cover the two most commonly used tools/submodules in HiTAD:</p>
<ul>
<li><p class="first">hitad</p>
<p>A command-line tool streamlining our 5-step identification algorithm:</p>
<ol class="arabic simple">
<li>Calculate adaptive directionality index (DI) for each bin.</li>
<li>Detect bottom boundaries by 5-state Gaussian mixture Hidden Markov Model using adaptive
DIs as input.</li>
<li>Identify TAD from bottom domains by optimally separating intra-chromosomal interactions
under chromosome-level objective functions.</li>
<li>Recursively identify inner domains under domain-level objective functions to optimally
separate intra-domain interactions.</li>
<li>(Optional but recommended) Perform domain alignment between two replicates and only
the reproducible domains are maintained to generate the final hierarchical TAD list.</li>
</ol>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If the two replicates are comparable in sequencing depth, then step 5 can improve
the accuracy and reliability while guaranteeing the sensitivity of called domains; otherwise,
you’d better merge the replicates into one dataset and step 5 will be skipped in this case.</p>
</div>
</li>
<li><p class="first">aligner</p>
<p>A submodule containing classes and functions for our proposed domain alignment strategy. In
our work, this strategy is used in reproducible domain detecting and change type defining
at both boundary and domain level.</p>
</li>
</ul>
</div>
<div class="section" id="tutorial">
<h2>Tutorial<a class="headerlink" href="#tutorial" title="Permalink to this headline">¶</a></h2>
<div class="section" id="hierarchical-tad-identification">
<h3>Hierarchical TAD Identification<a class="headerlink" href="#hierarchical-tad-identification" title="Permalink to this headline">¶</a></h3>
<div class="section" id="data-preparation">
<h4>Data Preparation<a class="headerlink" href="#data-preparation" title="Permalink to this headline">¶</a></h4>
<p>Since version 0.4.0, the input Hi-C matrix must be in <a class="reference external" href="https://github.com/mirnylab/cooler">cool</a> format.</p>
<p>Depending on what data you already have, there are different tools you can choose to generate
<em>cool</em>:</p>
<ul class="simple">
<li>If you are starting from the beginning (FASTQ/SRA), I recommend using <a class="reference external" href="https://github.com/XiaoTaoWang/HiC_pipeline">runHiC</a>,
a user-friendly and efficient Hi-C data processing tool developed by our lab.</li>
<li>If you are an old user of TADLib and have NPZ/TXT Hi-C matrix at hand, you can use the <em>toCooler</em>
script distributed with another software of mine <a class="reference external" href="https://github.com/XiaoTaoWang/HiCPeaks">hicpeaks</a>.</li>
<li>In other case, try <a class="reference external" href="https://cooler.readthedocs.io/en/latest/cli.html">cooler official tools</a>.</li>
</ul>
</div>
<div class="section" id="usage">
<h4>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">hitad</span> <span class="pre">[options]</span></code></p>
<p>To run hitad, you need to prepare a TXT meta data file specifying the resolution (in base-pair unit),
replicate label and cool URI information in a structured way. For example, if you want to use two
replicates to improve reliability of identified domains (as we did in our work), your metadata file
should look like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">res</span><span class="p">:</span><span class="nb">int</span>
  <span class="n">rep1</span><span class="p">:</span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">the</span><span class="o">/</span><span class="n">rep1</span><span class="o">/</span><span class="n">cool</span><span class="o">/</span><span class="n">URI</span>
  <span class="n">rep2</span><span class="p">:</span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">the</span><span class="o">/</span><span class="n">rep2</span><span class="o">/</span><span class="n">cool</span><span class="o">/</span><span class="n">URI</span>
</pre></div>
</div>
<p>Or if you have only one replicate or the sequencing depths between two replicates are quite different:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">res</span><span class="p">:</span><span class="nb">int</span>
  <span class="n">rep1</span><span class="p">:</span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">the</span><span class="o">/</span><span class="n">rep1</span><span class="o">/</span><span class="n">cool</span><span class="o">/</span><span class="n">URI</span>
</pre></div>
</div>
<p>Then just follow the pseudo command below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ hitad -O test.txt -d meta_file --logFile hitad.log
</pre></div>
</div>
<ul>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-O/--output</span></code> OUTPUT</p>
<p>Output file name.</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-d/--datasets</span></code> DATASETS</p>
<p>Metadata file path describing your Hi-C data</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">--logFile</span></code> LOGFILE</p>
<p>Log file name. (Default: hitad.log)</p>
</li>
</ul>
<p>Two files “test.txt” and “hitad.log” will be generated under current working directory.
We use a rotating file for logging. According to our settings, when the size of “hitad.log”
gets about 200K, it’s closed and renamed to “hitad.log.1”. At the same time, a new file
“hitad.log” is silently opened for output. In a word, the system saves old log files by
appending the extensions “.1”, “.2” etc., and the current log is always written to “hitad.log”:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="s2">&quot;test.txt&quot;</span> <span class="n">contains</span> <span class="mi">4</span> <span class="n">columns</span> <span class="n">indicating</span> <span class="n">chromosome</span> <span class="n">label</span><span class="p">,</span> <span class="n">domain</span> <span class="n">start</span> <span class="p">(</span><span class="n">bp</span><span class="p">),</span> <span class="n">domain</span> <span class="n">end</span> <span class="p">(</span><span class="n">bp</span><span class="p">),</span>
</pre></div>
</div>
<p>and hierarchical label, respectively. In our notation, TAD is denoted as level 0, sub-TAD is denoted
as level 1, and subsequent domain level is denoted as level 2, etc.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">22</span>      <span class="mi">18240000</span>        <span class="mi">18560000</span>        <span class="mi">0</span>
<span class="mi">22</span>      <span class="mi">19000000</span>        <span class="mi">19640000</span>        <span class="mi">0</span>
<span class="mi">22</span>      <span class="mi">19920000</span>        <span class="mi">20120000</span>        <span class="mi">0</span>
<span class="mi">22</span>      <span class="mi">20800000</span>        <span class="mi">21480000</span>        <span class="mi">0</span>
<span class="mi">22</span>      <span class="mi">22040000</span>        <span class="mi">22320000</span>        <span class="mi">0</span>
<span class="mi">22</span>      <span class="mi">22320000</span>        <span class="mi">22720000</span>        <span class="mi">1</span>
</pre></div>
</div>
</div>
<div class="section" id="other-options">
<h4>Other Options<a class="headerlink" href="#other-options" title="Permalink to this headline">¶</a></h4>
<ul>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">--maxsize</span></code> MAXSIZE</p>
<p>Maximum domain size in base-pair unit. (Default: 4000000)</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">--removeCache</span></code></p>
<p>Remove cached data before existing. (Default: False)</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-p/--cpu-core</span></code> CPU_CORE</p>
<p>Number of processes to launch. (Default: 1)</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-v/--version</span></code></p>
<p>Print version number and exit.</p>
</li>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">-h/--help</span></code></p>
<p>Show help message and exit.</p>
</li>
</ul>
</div>
</div>
<div class="section" id="visualization">
<h3>Visualization<a class="headerlink" href="#visualization" title="Permalink to this headline">¶</a></h3>
<p>Since version 0.4.0, I have included a handy script called <em>tad-plot</em> for hierarchical domain
visualization:</p>
<img alt="_images/region1.png" class="align-center" src="_images/region1.png" />
<p>On top of the heatmap is the adaptive DI track, which has been recorded in your <em>cool</em> file
after executing <em>hitad</em>. You can also choose to visualize domains on the arrowhead-transformed
heatmap:</p>
<img alt="_images/region1-arrowhead.png" class="align-center" src="_images/region1-arrowhead.png" />
</div>
<div class="section" id="output-di-track">
<h3>Output DI track<a class="headerlink" href="#output-di-track" title="Permalink to this headline">¶</a></h3>
<p>To facilitate integrative analysis and visualization in a standard genome browser, I also
added a script called <em>output-DI</em> in 0.4.0, which is able to export the adaptive DI track
in BedGraph format for any genomic regions.</p>
</div>
<div class="section" id="domain-alignment">
<h3>Domain Alignment<a class="headerlink" href="#domain-alignment" title="Permalink to this headline">¶</a></h3>
<p>Traditionally, domains and boundaries are aligned by matching boundaries with nearest ones
between two data sets. This strategy generally assign a threshold in advance to determine
whether a boundary of one data set exists in the other data set. However, the selection
of the threshold is quite empirical and artificial. To deal with this problem, we propose
a parameter-free alignment strategy by maximizing the overlap ratio between matched domains
and considering all domains at the same time. We also generalize the strategy on hierarchical
TADs and further define change types at both boundary and domain levels systematically.</p>
<p>All related classes and functions are defined in <code class="xref py py-mod docutils literal notranslate"><span class="pre">tadlib.hitad.aligner</span></code>, refer to
<a class="reference external" href="https://xiaotaowang.github.io/TADLib/hitad.html#api-documentation">API Documentation</a> for
more details.</p>
</div>
</div>
<div class="section" id="id1">
<h2>API Documentation<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h2>
<p>API reference of our defined classes and functions for hierarchical TAD
identification and domain alignment.</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="hitad_api.html">Hierarchical TAD Identification</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="hitad_api.html#domain-loading-and-aligning">Domain Loading and Aligning</a><ul class="simple">
</ul>
</li>
</ul>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2018, Xiaotao Wang.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/hitad.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/XiaoTaoWang/TADLib" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>