
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>TADLib &#8212; TADLib 0.5.0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Install TADLib" href="install.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="#">TADLib</a></h1>



<p class="blurb">A Library to Explore Chromatin Interaction Patterns for Topologically Associating Domains</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=XiaoTaoWang&repo=TADLib&type=star&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Install TADLib</a></li>
<li class="toctree-l1"><a class="reference internal" href="calfea.html">Aggregation Preference</a></li>
<li class="toctree-l1"><a class="reference internal" href="hitad.html">Hierarchical TAD</a></li>
<li class="toctree-l1"><a class="reference internal" href="domaincaller.html">Domain Caller</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Release Notes</a></li>
</ul>


<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="tadlib">
<h1>TADLib<a class="headerlink" href="#tadlib" title="Permalink to this headline">¶</a></h1>
<p>Chromosome conformation capture (3C) derived techniques, especially Hi-C, have
revealed that topologically associating domain (TAD) is a structural basis for
both chromatin organization and biological functions in three-dimensional (3D)
space. TAD is also hierarchically organized by smaller structural units, which
are relevant to biological functions. To systematically investigate the relationship
between structure and function, it is important to develop quantitative methods to
identify and measure the organization of TAD. TADLib is such a library to explore
the chromatin interaction patterns inside TAD from Hi-C chromatin interactions.</p>
<p>Currently, TADLib consists of three methods:</p>
<ul class="simple">
<li><dl class="first docutils">
<dt>Aggregation Preference (AP)</dt>
<dd>AP is a quantitative parameter to measure the overall density of significant
chromatin interactions inside TAD. Inspired by the observation that there exist
great differences in chromatin interaction pattern among TADs, an empirical
parameter called Aggregation Preference (AP) can be used to capture these
aggregation degree of significant chromatin interactions. Application to human
and mouse cell types (including both traditional Hi-C and in situ Hi-C data sets)
shows that there exist heterogeneous structures among TADs and the structural
rearrangement across cell types is significantly associated with transcriptional
remodelling. <a class="footnote-reference" href="#id4" id="id1">[1]</a></dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Hierarchical TAD (HiTAD)</dt>
<dd>HiTAD is a method to detect hierarchical TADs, including TADs, sub-TADs and
smaller domains. Except local insulations, HiTAD further constrains TADs as the
optimal domains to globally separate intra-chromosomal interactions. Under
objective functions derived from chromatin interactions, HiTAD adopts an iterative
optimization procedure to detect hierarchical TADs. HiTAD performs well in domain
sensitivity, replicate reproducibility and inter cell-type conservation. Application
to human and mouse cell types (including both traditional Hi-C and in situ Hi-C data
sets) reveals that there exist common change types for hierarchical TADs, which are
involved in shaping higher-order compartment, replication timing and transcriptional
regulation. <a class="footnote-reference" href="#id5" id="id2">[2]</a></dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Domain Caller (domaincaller)</dt>
<dd>Domaincaller is an implementation of the original Directionality Index (DI) based
TAD caller proposed by Dixon et al. <a class="footnote-reference" href="#id6" id="id3">[3]</a> Instead of original separate scripts for
each stage of the caller, this module provides a convenient command line interface
integrating the whole pipeline, including calculating the DI track, performing HMM
and post-processing. It supports the <a class="reference external" href="https://github.com/mirnylab/cooler">.cool</a>
matrix format, so has low memory requirements when dealing with high resolution data.
The TAD calls on the <a class="reference external" href="http://chromosome.sdsc.edu/mouse/hi-c/hESC.tar.gz">Human ES dataset</a>
are pretty similar to the published list, both in boundary locations and CTCF enrichment.</dd>
</dl>
</li>
</ul>
<div class="section" id="user-guide">
<h2>User Guide<a class="headerlink" href="#user-guide" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Install TADLib</a><ul>
<li class="toctree-l2"><a class="reference internal" href="install.html#requirements">Requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="install.html#install-conda">Install Conda</a></li>
<li class="toctree-l2"><a class="reference internal" href="install.html#install-the-required-packages">Install the Required Packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="install.html#id1">Install TADLib</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="calfea.html">Aggregation Preference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="calfea.html#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="calfea.html#quickstart">QuickStart</a><ul>
<li class="toctree-l3"><a class="reference internal" href="calfea.html#usage">Usage</a></li>
<li class="toctree-l3"><a class="reference internal" href="calfea.html#other-options">Other Options</a></li>
<li class="toctree-l3"><a class="reference internal" href="calfea.html#next-steps">Next Steps</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="calfea.html#api-documentation">API Documentation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="calfea_api.html">TAD Data Parsing</a></li>
<li class="toctree-l3"><a class="reference internal" href="calfea_api.html#intra-tad-interaction-analysis">Intra-TAD Interaction Analysis</a></li>
<li class="toctree-l3"><a class="reference internal" href="calfea_api.html#matrix-manipulating">Matrix Manipulating</a></li>
<li class="toctree-l3"><a class="reference internal" href="calfea_api.html#polygon-creation-and-operations">Polygon Creation and Operations</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="hitad.html">Hierarchical TAD</a><ul>
<li class="toctree-l2"><a class="reference internal" href="hitad.html#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="hitad.html#snapshot">Snapshot</a></li>
<li class="toctree-l2"><a class="reference internal" href="hitad.html#tutorial">Tutorial</a><ul>
<li class="toctree-l3"><a class="reference internal" href="hitad.html#hierarchical-tad-identification">Hierarchical TAD Identification</a></li>
<li class="toctree-l3"><a class="reference internal" href="hitad.html#visualization">Visualization</a></li>
<li class="toctree-l3"><a class="reference internal" href="hitad.html#output-di-track">Output DI track</a></li>
<li class="toctree-l3"><a class="reference internal" href="hitad.html#domain-alignment">Domain Alignment</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="hitad.html#id1">API Documentation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="hitad_api.html">Hierarchical TAD Identification</a></li>
<li class="toctree-l3"><a class="reference internal" href="hitad_api.html#domain-loading-and-aligning">Domain Loading and Aligning</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="domaincaller.html">Domain Caller</a><ul>
<li class="toctree-l2"><a class="reference internal" href="domaincaller.html#tutorial">Tutorial</a><ul>
<li class="toctree-l3"><a class="reference internal" href="domaincaller.html#data-preparation">Data Preparation</a></li>
<li class="toctree-l3"><a class="reference internal" href="domaincaller.html#usage">Usage</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="domaincaller.html#performance-test">Performance test</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Release Notes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#version-0-5-0-2019-05-03">Version 0.5.0 (2019-05-03)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#version-0-4-0-2018-11-24">Version 0.4.0 (2018-11-24)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#version-0-3-1-r1-2017-8-18">Version 0.3.1-r1 (2017-8-18)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#version-0-3-1-2017-7-19">Version 0.3.1 (2017-7-19)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#version-0-3-0-dev2-2017-5-28">Version 0.3.0-dev2 (2017-5-28)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#version-0-3-0-dev1-2017-5-2">Version 0.3.0-dev1 (2017-5-2)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#version-0-2-5-r2-2016-4-14">Version 0.2.5-r2 (2016-4-14)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#version-0-2-5-r1-2015-11-10">Version 0.2.5-r1 (2015-11-10)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#version-0-2-5-2015-11-7">Version 0.2.5 (2015-11-7)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#version-0-2-4-2015-4-14">Version 0.2.4 (2015-4-14)</a></li>
<li class="toctree-l2"><a class="reference internal" href="changelog.html#version-0-1-0-2014-11-25">Version 0.1.0 (2014-11-25)</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="reference">
<h2>Reference<a class="headerlink" href="#reference" title="Permalink to this headline">¶</a></h2>
<table class="docutils footnote" frame="void" id="id4" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[1]</a></td><td>Wang XT, Dong PF, Zhang HY, Peng C. Structural heterogeneity and functional diversity
of topologically associating domains in mammalian genomes. Nucleic Acids Research, 2015,
doi: 10.1093/nar/gkv684</td></tr>
</tbody>
</table>
<table class="docutils footnote" frame="void" id="id5" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id2">[2]</a></td><td>Wang XT, Cui W, Peng C. HiTAD: detecting the structural and functional hierarchies of
topologically associating domains from chromatin interactions. Nucleic Acids Research, 2017,
doi: 10.1093/nar/gkx735</td></tr>
</tbody>
</table>
<table class="docutils footnote" frame="void" id="id6" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id3">[3]</a></td><td>Dixon JR, Selvaraj S, Yue F, Kim A, Li Y, Shen Y, Hu M, Liu JS, Ren B. Topological domains
in mammalian genomes identified by analysis of chromatin interactions. Nature, 2012,
doi: 10.1038/nature11082</td></tr>
</tbody>
</table>
</div>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2017, XiaoTao Wang.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/XiaoTaoWang/TADLib" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>